### PR TITLE
Lower `linearized` into a series of adds and muls.

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue id = #464 

I'm confused with those `Stmt*`'s and `std::unique_ptr`'s. The code cannot pass the test now and I'm not sure if I'm writing bugs.

### CODE REVIEW

**Comment by yuanming-hu** [comment]
One tip: `unique_ptr` means ownership yet raw pointers do not. Google `C++ smart pointer ownership` to learn more about it.

**Comment by xumingkuan** [comment]
Just to confirm that the product of `stmt->strides[i]` will not exceed `INT_MAX`, right?

**Comment by yuanming-hu** [comment]
> Just to confirm that the product of stmt->strides[i] will not exceed INT_MAX, right?

Good question! Yes, you can assume it fits in `int32`.

File: taichi/transforms/simplify.cpp [hunk 1]
```diff
@@ -661,9 +661,6 @@ class BasicBlockSimplify : public IRVisitor {
   }
 
   void visit(LinearizeStmt *stmt) override {
-    if (is_done(stmt))
-      return;
-
     if (stmt->inputs.size() && stmt->inputs.back()->is<IntegerOffsetStmt>()) {
```
Comment (yuanming-hu) [inline taichi/transforms/simplify.cpp]: Let's comment out this clause. I need to think a bit more about if this part is necessary.

File: taichi/transforms/simplify.cpp [hunk 2]
```diff
@@ -691,7 +688,32 @@ class BasicBlockSimplify : public IRVisitor {
         }
```
Comment (yuanming-hu) [inline taichi/transforms/simplify.cpp]: The CSE here 
```
    for (int i = 0; i < current_stmt_id; i++) {
      auto &bstmt = block->statements[i];
      if (stmt->ret_type == bstmt->ret_type) {
        auto &bstmt_data = *bstmt;
        if (typeid(bstmt_data) == typeid(*stmt)) {
          auto bstmt_ = bstmt->as<LinearizeStmt>();
          if (identical_vectors(bstmt_->inputs, stmt->inputs) &&
              identical_vectors(bstmt_->strides, stmt->strides)) {
            stmt->replace_with(bstmt.get());
            stmt->parent->erase(current_stmt_id);
            throw IRModified();
          }
        }
      }
    }
```

can be removed now.

File: taichi/transforms/simplify.cpp [hunk 3]
```diff
@@ -691,7 +688,32 @@ class BasicBlockSimplify : public IRVisitor {
...
+        auto stride_stmt = Stmt::make<ConstStmt>(LaneAttribute<TypedConstant>(stride_product));
+        auto mul = Stmt::make<BinaryOpStmt>(BinaryOpType::mul, stmt->inputs[i], stride_stmt.get());
+        auto sum = Stmt::make<BinaryOpStmt>(BinaryOpType::add, sumptr, mul.get());
+        sumptr = sum.get();
+        stmt->insert_before_me(std::move(stride_stmt));
+        stmt->insert_before_me(std::move(mul));
+        stmt->insert_before_me(std::move(sum));
+      }
+      stmt->replace_with(sumptr);
+      stmt->parent->erase(stmt);
```
Comment (yuanming-hu) [inline taichi/transforms/simplify.cpp]: You can do `irpass::typecheck(stmt->parent)` if necessary.

File: taichi/transforms/simplify.cpp [hunk 4]
```diff
@@ -691,7 +688,32 @@ class BasicBlockSimplify : public IRVisitor {
         }
       }
     }
-    set_done(stmt);
+
+    // Lower into a series of adds and muls.
+    // Need typecheck afterwards.
+    if (stmt->inputs.empty()) {
```
Comment (yuanming-hu) [inline taichi/transforms/simplify.cpp]: Actually, no need to specialize when you have empty inputs. Empty inputs should simply be treated as `inputs` with zero lengths. Then you can merge two clauses of this `if` stmt. The extra `0 + i` can be optimized by your `alg_simp` pass.

File: tests/cpp/test_alg_simp.cpp [hunk 5]
```diff
@@ -121,4 +121,39 @@ TI_TEST("simplify_multiply_zero_fast_math") {
   TI_CHECK((*block)[0]->is<GlobalTemporaryStmt>());
 }
 
+TI_TEST("simplify_linearized_with_trivial_inputs") {
```
Comment (yuanming-hu) [inline tests/cpp/test_alg_simp.cpp]: Maybe create a `test_simplify.cpp` file for this test?

**Comment by yuanming-hu** [comment]
Looks good in general! Just some small changes are needed.

File: taichi/transforms/simplify.cpp (hunk 1 above)
Comment (xumingkuan) [reply taichi/transforms/simplify.cpp]: I think `stmt` no longer exists once `set_done(stmt);` and the following code are executed. Added the comment back anyway.

Comment (yuanming-hu) [reply taichi/transforms/simplify.cpp]: Sounds good!

**Comment by xumingkuan** [comment]
Changes finished :)

**Pull request merged**